RPC parsing fixes

mmakassikis opened the pull request:
It is possible to crash ksmbd.mountd by sending invalid RPC requests.

This series adds some missing checks that fix crashes turned up by fuzzing the RPC code. The last commit is quite large as callers to ndr_write_* functions are modifed to check the return value and exit early. This is a mechanical change, but it's possible I missed something in the process.

namjaejeon commented on mountd/rpc.c:
```diff
@@ -1002,9 +1116,17 @@ static int dcerpc_bind_ack_return(struct ksmbd_rpc_pipe *pipe)
-	ndr_write_int16(dce, dce->bi_req.max_xmit_frag_sz);
-	ndr_write_int16(dce, dce->bi_req.max_recv_frag_sz);
-	ndr_write_int32(dce, dce->bi_req.assoc_group_id);
+	ret = ndr_write_int16(dce, dce->bi_req.max_xmit_frag_sz);
```
We need ret variable ? how is this ?
```
if (ndr_write_int16(dce, dce->bi_req.max_xmit_frag_sz))
     return KSMBD_RPC_EBAD_DATA;
```

mmakassikis replied:
I started with having ret variable and then copied the same pattern pretty much everywhere. Do you want me to change the occurrences to what you suggest instead ?

namjaejeon commented on mountd/rpc.c:
```diff
@@ -325,7 +325,9 @@ int ndr_write_union_##name(struct ksmbd_dcerpc *dce, type value)	\
 	ret = ndr_write_##name(dce, value);				\
-	ret |= ndr_write_##name(dce, value);				\
+	if (ret)							\
+		return -ENOMEM;						\
```
any reason why you return -ENOMEM here instead of ret variable ?

mmakassikis replied:
Nope, no reason

mmakassikis commented:
I pushed an updated version:
 - fix invalid free in "ksmbd-tools: mountd: validate RPC request"
 - additional fixes for set_domain_name()

namjaejeon commented on mountd/rpc_lsarpc.c:
```diff
@@ -760,7 +760,9 @@ int rpc_lsarpc_init(void)
 	 * uses the hostname as the domain name.
 	 */
-	gethostname(domain_string, NAME_MAX);
+	if (gethostname(domain_string, NAME_MAX))
```
we don't check return value of rpc_samr_init() and rpc_lsarpc_init(). I think that we need to add error handling for them.

namjaejeon commented on mountd/smbacl.c:
```diff
@@ -188,12 +188,20 @@ int set_domain_name(struct smb_sid *sid, char *domain, int *type)
+		if (ret > domain_len)
+			return -ENOMEM;
+
+		domain_name = g_ascii_strup(domain_string, -1);
```
we need to free domain_name at the end of this function ?

namjaejeon commented on mountd/smbacl.c:
```diff
@@ -188,12 +188,20 @@ int set_domain_name(struct smb_sid *sid, char *domain, int *type)
-		domain_name = g_ascii_strup(domain_string,
-				strlen(domain_string));
+		ret = smb_sid_to_string(domain_string, sizeof(domain_string),
+					sid);
```
we don't need same check the above strcpy() ?

namjaejeon commented on include/rpc_lsarpc.h:
```diff
@@ -11,6 +11,7 @@
 #include <smbacl.h>
 
 #define HANDLE_SIZE	20
+#define DOMAIN_STR_SIZE	(NAME_MAX + 2)
```
I think that we can declare 257 instead of NAME_MAX.

mmakassikis commented on mountd/worker.c:
```diff
@@ -220,10 +220,14 @@ static int rpc_request(struct ksmbd_ipc_msg *msg)
+		goto out;
+
 	req = KSMBD_IPC_MSG_PAYLOAD(msg);
+	req->payload_sz = msg->sz - sizeof(struct ksmbd_rpc_command);
```
Looking at this again, I don't think this line is necessary. Without it, I had crashes while fuzzing, but that's because my code was wrong to begin with (it only set req->sz, without touching req->payload_sz).

mmakassikis commented:
Updated the branch:
 - replaced strcpy() with snprintf() in set_domain_name(), and updated snprintf error checking (previously, we only checked for overflow)
 - removed req->payload_sz overwriting
 - minimized the use of intermediate 'ret' variable when checking ndr_write_* return
 - added some more ndr_write_* checks that were missing
 - added check for rpc_samr_init() / rpc_lsarpc_init() in rpc_init()

namjaejeon commented:
Applied. Thanks for your patches!
